pallets/pooled-staking/src/candidate.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+use {
+    crate::{
+        pools::{self, Pool},
+        traits::{ErrAdd, ErrSub},
+        Candidate, CollatorSet, Config, Error, Event, MaxCollatorSetSize, Pallet, Pools, PoolsKey,
+        SortedEligibleCandidates, Stake,
+    },
+    core::{cmp::Ordering, marker::PhantomData},
+    parity_scale_codec::{Decode, Encode},
+    scale_info::TypeInfo,
+    sp_core::{Get, RuntimeDebug},
+    sp_runtime::traits::Zero,
+    sp_std::collections::btree_set::BTreeSet,
+};
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+/// Eligible candidate with its stake.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo)]
+pub struct EligibleCandidate<C, S> {
+    candidate: C,
+    stake: S,
+}
+
+impl<C: Ord, S: Ord> Ord for EligibleCandidate<C, S> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.stake
+            .cmp(&other.stake)
+            .reverse()
+            .then(self.candidate.cmp(&self.candidate))
+    }
+}
+
+impl<C: Ord, S: Ord> PartialOrd for EligibleCandidate<C, S> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub struct Candidates<T>(PhantomData<T>);
+
+impl<T: Config> Candidates<T> {
+    pub fn total_stake(candidate: &Candidate<T>) -> Stake<T> {
+        Stake(Pools::<T>::get(candidate, &PoolsKey::CandidateTotalStake))
+    }
+
+    pub fn add_total_stake(candidate: &Candidate<T>, stake: &Stake<T>) -> Result<(), Error<T>> {
+        if Zero::is_zero(&stake.0) {
```

Comment:
maybe we can implement the Zero trait to stake?

Reply:
I would avoid implementing all arithmetic traits on those. Their goal is just to be wrappers to avoid confusion between amounts in currency or shares.